README.md
# idkit_line

This is a convenient development kit for the expansion of the requirements of the dividing line in the development of the Flutter project.

## Use case of the package extension method

- Solid line

  ```dart
  IDKitLine.solid(),
  const SizedBox(height: 20),
  IDKitLine.solid(
      color: Colors.red,
      width: 300,
      height: 5,
      thickness: 3,
  ),
  const SizedBox(height: 20),
  IDKitLine.solid(
      color: Colors.purple,
      width: 300,
      axis: Axis.vertical,
      height: 20,
  )
  ```

- Dotted line
  ```dart
  IDKitLine.dotted(),
  const SizedBox(height: 20),
  IDKitLine.dotted(
      dashLength: 5,
      interval: 3,
      thickness: 3,
  ),
  const SizedBox(height: 20),
  IDKitLine.dotted(
      width: 200,
      color: Colors.purple,
      dashLength: 5,
      interval: 3,
  ),
  const SizedBox(height: 20),
  IDKitLine.dotted(
      dashLength: 5,
      interval: 3,
      dottedType: DottedType.dash_dot,
  ),
  const SizedBox(height: 20),
  IDKitLine.dotted(
      width: 200,
      color: Colors.purple,
      dashLength: 10,
      interval: 3,
      thickness: 4,
      dottedType: DottedType.dash_dot,
  )
  ```
- Wavy line
  ```dart
  IDKitLine.wavy(),
  const SizedBox(height: 20),
  IDKitLine.wavy(
      color: Colors.red,
      width: 300,
  ),
  const SizedBox(height: 20),
  IDKitLine.wavy(
      axis: Axis.vertical,
      height: 200,
      thickness: 4,
      a: 6,
      w: 0.2,
  )
  ```
- Delete line
  ```dart
  IDKitLine.delete(
    child: const Text('¥：10'),
    thickness: 2,
    color: Colors.red,
  ),
  const SizedBox(height: 20),
  IDKitLine.delete(
    child: const Text('¥：10000000000'),
    thickness: 4,
    color: Colors.purple,
  ),
  ```

## Point

If you want to know more about idkitline usage, please go to Github and search idkitline,[Click here for quick access](https://github.com/zhoushuangjian001/idkit_line).


CHANGELOG.md
## 0.0.1

This is a convenient development kit for the expansion of the requirements of the dividing line in the development of the Flutter project.

#### Extension of the package drawing method

- Solid line

  ```dart
  /// Factory method of solid line
  factory IDKitLine.solid({...}) =>IDKitLine;
  ```

- Wavy line
  ```dart
  /// Wavy line factory method
  factory IDKitLine.wavy({...}) =>IDKitLine;
  ```
- Dotted line
  ```dart
  /// Dotted factory method
  factory IDKitLine.dotted({...}) =>IDKitLine;
  ```
- Delete lin
  ```dart
  /// Delete the factory initialization method of the line
  factory IDKitLine.delete({...}) =>IDKitLine;
  ```

#### Hint

For more usage, please look [README.md](https://github.com/zhoushuangjian001/idkit_line#readme)


test/idkit_line_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('Extension test of line package', () {});
}


lib/src/line_main.dart
import 'package:flutter/material.dart'
    show StatelessWidget, Axis, Key, Color, LayoutBuilder, BuildContext, BoxConstraints, SizedBox, CustomPaint, Size, Colors, Widget;
import 'package:idkit_line/idkit_line.dart';
import 'package:idkit_line/src/line_config.dart';
import 'package:idkit_line/src/line_painter.dart';

class IDKitLine extends StatelessWidget {
  const IDKitLine({
    Key? key,
    this.width,
    this.height,
    this.type = LineType.solid,
    this.thickness,
    this.axis = Axis.horizontal,
    this.align = LineAlign.center,
    this.color,
    this.indent,
    this.endIndent,
    this.a,
    this.w,
    this.k,
    this.dottedType = DottedType.dash,
    this.dashLength,
    this.interval,
    this.child,
  })  : assert(!(axis == Axis.vertical && height == null), '\nPoint:Draw a line in the vertical direction, the height must be set.'),
        super(key: key);

  /// Factory method for drawing strikethrough
  factory IDKitLine.delete({
    Key? key,
    double? thickness,
    Color? color,
    Widget? child,
  }) =>
      IDKitLine(
        key: key,
        thickness: thickness,
        color: color,
        child: child,
      );

  /// Factory method of solid line
  factory IDKitLine.solid({
    Key? key,
    double? width,
    double? height,
    double? thickness,
    Axis axis = Axis.horizontal,
    LineAlign align = LineAlign.center,
    Color? color,
    double? indent,
    double? endIndent,
  }) =>
      IDKitLine(
        key: key,
        width: width,
        height: height,
        thickness: thickness,
        axis: axis,
        align: align,
        color: color,
        indent: indent,
        endIndent: endIndent,
      );

  /// Wavy line factory method
  factory IDKitLine.wavy({
    Key? key,
    double? width,
    double? height,
    LineType type = LineType.wavy,
    double? thickness,
    Axis axis = Axis.horizontal,
    LineAlign align = LineAlign.center,
    Color? color,
    double? indent,
    double? endIndent,
    double? a,
    double? w,
    double? k,
  }) =>
      IDKitLine(
        key: key,
        width: width,
        height: height,
        type: type,
        thickness: thickness,
        axis: axis,
        align: align,
        color: color,
        indent: indent,
        endIndent: endIndent,
        a: a,
        w: w,
        k: k,
      );

  /// Dotted factory method
  factory IDKitLine.dotted({
    Key? key,
    double? width,
    double? height,
    LineType type = LineType.dotted,
    double? thickness,
    Axis axis = Axis.horizontal,
    LineAlign align = LineAlign.center,
    Color? color,
    double? indent,
    double? endIndent,
    DottedType dottedType = DottedType.dash,
    double? dashLength,
    double? interval,
  }) =>
      IDKitLine(
        key: key,
        width: width,
        height: height,
        type: type,
        thickness: thickness,
        axis: axis,
        align: align,
        color: color,
        indent: indent,
        endIndent: endIndent,
        dottedType: dottedType,
        dashLength: dashLength,
        interval: interval,
      );

  /// The width of the carrier to draw the line
  final double? width;

  /// The height of the carrier to draw the line
  final double? height;

  /// Type of drawing line
  final LineType type;

  /// The thickness of the drawn line
  final double? thickness;

  /// The direction of the drawn line
  final Axis axis;

  /// The position of the drawn line
  final LineAlign align;

  /// The color of the drawn line
  final Color? color;

  /// Start indentation of the line drawn
  final double? indent;

  /// End indentation of the line drawn
  final double? endIndent;

  /// Draw the amplitude of the wave line
  final double? a;

  /// Angular velocity for drawing wavy lines
  final double? w;

  /// Draw the offset of the wavy line
  final double? k;

  /// Type of dashed line
  final DottedType dottedType;

  /// The length of the dashed line
  final double? dashLength;

  /// Space between dotted elements
  final double? interval;

  final Widget? child;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: width,
      height: height,
      child: LayoutBuilder(
        builder: (BuildContext context, BoxConstraints constraints) {
          Size painterSize = Size.zero;
          if (constraints.isTight) {
            painterSize = Size(constraints.minWidth, constraints.minHeight);
          } else if (constraints.isNormalized) {
            painterSize = Size(constraints.maxWidth, constraints.maxHeight == double.infinity ? 1 : constraints.maxHeight);
          }
          final bool isForeground = type == LineType.delete;
          return CustomPaint(
            size: painterSize,
            foregroundPainter: isForeground
                ? LinePainter(
                    context,
                    type: LineType.solid,
                    axis: Axis.horizontal,
                    align: LineAlign.center,
                    thickness: thickness,
                    color: color,
                  )
                : null,
            painter: isForeground
                ? null
                : LinePainter(
                    context,
                    type: type,
                    axis: axis,
                    align: align,
                    thickness: thickness,
                    color: color,
                    a: a,
                    w: w,
                    k: k,
                    dottedType: dottedType,
                    dashLength: dashLength,
                    interval: interval,
                  ),
            child: child,
          );
        },
      ),
    );
  }
}


lib/src/line_mixin.dart
import 'dart:math' show sin;
import 'package:flutter/material.dart' show Canvas, Size, Axis, Paint, Offset, Path;
import 'package:idkit_line/src/line_config.dart';

mixin LinePainterMixin {
  /// Method of dashed line realization
  void drawDotted({
    required Canvas canvas,
    required Size size,
    required Paint paint,
    required double thickness,
    Axis? axis,
    LineAlign? align,
    double? indent,
    double? endIndent,
    DottedType? type,
    double? dashLength,
    double? interval,
  }) {
    // Default settings for constants
    final double _indent = indent ?? 0.0;
    final double _endIndent = endIndent ?? 0.0;
    final double totleIndent = _indent + _endIndent;
    final double halfThickness = thickness * 0.5;
    final bool isAxis = axis == Axis.horizontal;
    final double _dashLength = dashLength ?? 1;
    final double _interval = interval ?? 1;
    double _itemLength = 0.0;
    int multiple = 1;
    // Types of dashed lines
    switch (type) {
      case DottedType.dash_dot:
        // -- - -- -
        _itemLength = _dashLength + _interval * 2 + thickness;
        multiple = 2;
        break;
      default:
        // - - - -
        _itemLength = _dashLength + _interval;
    }
    // End error handling
    final double length = isAxis ? size.width : size.height;
    final double otherLength = isAxis ? size.height : size.width;
    final double lastLength = length - totleIndent - _dashLength;
    final int count = lastLength ~/ _itemLength;
    final double overLength = lastLength - count * _itemLength;
    final double _intervalDeviation = overLength / (multiple * count);
    _itemLength += multiple * _intervalDeviation;
    double _dx, _dy = 0;
    // Location distinction
    switch (align) {
      case LineAlign.top:
      case LineAlign.left:
        _dx = _indent;
        _dy = halfThickness;
        break;
      case LineAlign.bottom:
      case LineAlign.right:
        _dx = _indent;
        _dy = otherLength - halfThickness;
        break;
      default:
        _dx = _indent;
        _dy = otherLength * 0.5 - halfThickness;
    }
    // Start drawing
    Offset offset1, offset2;
    for (double i = 0; i < count; i++) {
      if (multiple == 2) {
        offset1 = Offset(_dx + i * _itemLength, _dy);
        offset2 = offset1.translate(_dashLength, 0);
        canvas.drawLine(isAxis ? offset1 : offset1.exchange(), isAxis ? offset2 : offset2.exchange(), paint);
        offset1 = offset2.translate(_interval, 0);
        offset2 = offset1.translate(thickness, 0);
        canvas.drawLine(isAxis ? offset1 : offset1.exchange(), isAxis ? offset2 : offset2.exchange(), paint);
      } else {
        offset1 = Offset(_dx + i * _itemLength, _dy);
        offset2 = offset1.translate(_dashLength, 0);
        canvas.drawLine(isAxis ? offset1 : offset1.exchange(), isAxis ? offset2 : offset2.exchange(), paint);
      }
    }
    final Offset offset3 = Offset(count * _itemLength + _dx, _dy);
    final Offset offset4 = offset3.translate(_dashLength, 0);
    canvas.drawLine(isAxis ? offset3 : offset3.exchange(), isAxis ? offset4 : offset4.exchange(), paint);
  }

  /// The method of solid line realization
  void drawSolid({
    required Canvas canvas,
    required Size size,
    required Paint paint,
    required double thickness,
    Axis? axis,
    LineAlign? align,
    double? indent,
    double? endIndent,
  }) {
    Offset p1 = Offset.zero;
    Offset p2 = Offset.zero;
    // Constant definition
    final double _indent = indent ?? 0.0;
    final double _endIndent = endIndent ?? 0.0;
    final double totleIndent = _indent + _endIndent;
    final double halfThickness = thickness * 0.5;
    final bool isAxis = axis == Axis.horizontal;
    final double totle = isAxis ? size.width : size.height;
    final double _totle = isAxis ? size.height : size.width;
    final double lineLength = totle - totleIndent;
    // Location distinction
    double offset = 0.0;
    switch (align) {
      case LineAlign.top:
      case LineAlign.left:
        offset = halfThickness;
        break;
      case LineAlign.bottom:
      case LineAlign.right:
        offset = _totle - halfThickness;
        break;
      default:
        final double half = (_totle - thickness) / 2;
        offset = half + halfThickness;
    }

    // Start drawing
    p1 = p1.translate(isAxis ? _indent : offset, isAxis ? offset : _indent);
    p2 = p2.translate(isAxis ? lineLength : offset, isAxis ? offset : lineLength);
    canvas.drawLine(p1, p2, paint);
  }

  /// Ways to realize wavy lines
  void drawWavy({
    required Canvas canvas,
    required Size size,
    required Paint paint,
    required double thickness,
    Axis? axis,
    LineAlign? align,
    double? indent,
    double? endIndent,
    double? a,
    double? k,
    double? w,
  }) {
    final double halfThickness = thickness * 0.5;
    // Total indent
    final double _indent = indent ?? 0.0;
    final double _endIndent = endIndent ?? 0.0;
    final double totleIndent = _indent + _endIndent;
    // Constant definition
    final bool isAxis = axis == Axis.horizontal;
    final double totle = isAxis ? size.width : size.height;
    final double _totle = isAxis ? size.height : size.width;
    final double wavyLength = totle - totleIndent;
    double lastOffset = 0.0;
    final double _a = a ?? 1;
    final double _w = w ?? 1;
    final double _k = k ?? 0;
    // Location distinction
    switch (align) {
      case LineAlign.top:
      case LineAlign.left:
        lastOffset = _a + halfThickness;
        break;
      case LineAlign.bottom:
      case LineAlign.right:
        lastOffset = _totle - (_a + halfThickness);
        break;
      default:
        lastOffset = _totle * 0.5;
    }
    // Start drawing
    final Path path = Path();
    path.moveTo(isAxis ? _indent : lastOffset, isAxis ? lastOffset : _indent);
    for (double x = _indent; x < wavyLength; x += 0.5) {
      final double offsetY = _a * sin(_w * x) + _k;
      final double y = offsetY + lastOffset;
      path.lineTo(isAxis ? x : y, isAxis ? y : x);
    }
    canvas.drawPath(path, paint);
  }
}


lib/src/line_painter.dart
import 'package:flutter/material.dart';
import 'package:idkit_line/src/line_config.dart';
import 'package:idkit_line/src/line_mixin.dart';

class LinePainter extends CustomPainter with LinePainterMixin {
  const LinePainter(
    this.context, {
    this.type,
    this.axis,
    this.align,
    this.dottedType,
    this.thickness,
    this.color,
    this.indent,
    this.endIndent,
    this.a,
    this.w,
    this.k,
    this.dashLength,
    this.interval,
  });
  final BuildContext context;
  final LineType? type;
  final Axis? axis;
  final double? thickness;
  final Color? color;
  final LineAlign? align;
  final double? indent;
  final double? endIndent;
  final double? a;
  final double? w;
  final double? k;
  final DottedType? dottedType;
  final double? dashLength;
  final double? interval;
  @override
  void paint(Canvas canvas, Size size) {
    /// 画笔设置
    final double _thickness = thickness ?? 1.0;
    final Paint paint = Paint();
    paint.style = PaintingStyle.stroke;
    paint.strokeWidth = _thickness;
    paint.color = color ?? DividerTheme.of(context).color ?? Theme.of(context).primaryColor;
    switch (type) {
      case LineType.dotted:
        drawDotted(
          canvas: canvas,
          size: size,
          paint: paint,
          thickness: _thickness,
          align: align,
          axis: axis,
          dashLength: dashLength,
          interval: interval,
          type: dottedType,
        );
        break;
      case LineType.wavy:
        drawWavy(
          canvas: canvas,
          size: size,
          paint: paint,
          indent: indent,
          axis: axis,
          align: align,
          thickness: _thickness,
          endIndent: endIndent,
          a: a,
          w: w,
          k: k,
        );
        break;
      default:
        drawSolid(
          canvas: canvas,
          size: size,
          paint: paint,
          axis: axis,
          align: align,
          indent: indent,
          endIndent: endIndent,
          thickness: _thickness,
        );
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => false;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:idkit_line/idkit_line.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Extension test of line package'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            IDKitLine.solid(),
            const SizedBox(height: 20),
            IDKitLine.solid(
              color: Colors.red,
              width: 300,
              height: 5,
              thickness: 3,
            ),
            const SizedBox(height: 20),
            IDKitLine.solid(
              color: Colors.purple,
              width: 300,
              axis: Axis.vertical,
              height: 20,
            ),
            const SizedBox(height: 20),
            IDKitLine.dotted(),
            const SizedBox(height: 20),
            IDKitLine.dotted(
              dashLength: 5,
              interval: 3,
              thickness: 3,
            ),
            const SizedBox(height: 20),
            IDKitLine.dotted(
              width: 200,
              color: Colors.purple,
              dashLength: 5,
              interval: 3,
            ),
            const SizedBox(height: 20),
            IDKitLine.dotted(
              dashLength: 5,
              interval: 3,
              dottedType: DottedType.dash_dot,
            ),
            const SizedBox(height: 20),
            IDKitLine.dotted(
              width: 200,
              color: Colors.purple,
              dashLength: 10,
              interval: 3,
              thickness: 4,
              dottedType: DottedType.dash_dot,
            ),
            const SizedBox(height: 20),
            Row(
              children: <Widget>[
                IDKitLine.dotted(
                  width: 200,
                  height: 30,
                  color: Colors.purple,
                  dashLength: 5,
                  interval: 3,
                  axis: Axis.vertical,
                ),
                IDKitLine.dotted(
                  width: 200,
                  height: 30,
                  color: Colors.purple,
                  dashLength: 5,
                  interval: 3,
                  dottedType: DottedType.dash_dot,
                  axis: Axis.vertical,
                )
              ],
            ),
            const SizedBox(height: 20),
            IDKitLine.wavy(),
            const SizedBox(height: 20),
            IDKitLine.wavy(
              color: Colors.red,
              width: 300,
            ),
            const SizedBox(height: 20),
            IDKitLine.wavy(
              axis: Axis.vertical,
              height: 200,
              thickness: 4,
              a: 6,
              w: 0.2,
            ),
            const SizedBox(height: 20),
            IDKitLine.delete(
              child: const Text('¥：10'),
              thickness: 2,
              color: Colors.red,
            ),
            const SizedBox(height: 20),
            IDKitLine.delete(
              child: const Text('¥：10000000000'),
              thickness: 4,
              color: Colors.purple,
            ),
          ],
        ),
      ),
    );
  }
}
